src/olympia/devhub/tasks.py

```diff
@@ -695,23 +631,63 @@ def run_validator(path, for_appversions=None, test_all_tiers=False,
             shutil.copyfileobj(storage.open(path), temp.file)
             path = temp.name
 
+        cmd = [
+            settings.ADDONS_VALIDATOR_BIN,
+            '--boring',
+            '--output=json',
+            '--approved_applications=' + apps,
+            '--timeout=' + str(settings.VALIDATOR_TIMEOUT),
+        ]
+
+        if not listed:
+            cmd.append('--selfhosted')
+
+        cmd.append(path)
+
+        stdout, stderr = (
+            tempfile.TemporaryFile(),
+            tempfile.TemporaryFile())
+
         with statsd.timer('devhub.validator'):
-            json_result = validator_validate(
-                path,
-                for_appversions=for_appversions,
-                format='json',
-                # When False, this flag says to stop testing after one
-                # tier fails.
-                determined=test_all_tiers,
-                approved_applications=apps,
-                overrides=overrides,
-                compat_test=compat,
-                listed=listed
+            process = subprocess.Popen(
+                # Have to use a string instead of a list because of how
+                # `shell=True` resolves paths and working dirs
+                ' '.join(cmd),
+                stdout=stdout,
+                stderr=stderr,
+                # We're explicitly using a shell here because amo-validator
+                # uses subprocess itself to shell out calls to spidermonkey
+                # and it somehow needs to be called with shell=True to make
+                # all that work properly. Not too much investigation went into
+                # the "why" but since `shell=True` isn't much of a security
+                # concern here given that all legacy add-ons we're validating
+                # are uploaded / developed by Mozilla employees at this point
+                # we're kinda safe-ish...
+                shell=True
             )
 
-        track_validation_stats(json_result)
+            process.wait()
 
-        return json_result
+            stdout.seek(0)
+            stderr.seek(0)
+
+            output, error = stdout.read(), stderr.read()
+
+            # Make sure we close all descriptors, otherwise they'll hang around
+            # and could cause a nasty exception.
+            stdout.close()
+            stderr.close()
+
+    if error:
+        raise ValueError(error)
+
+    parsed_data = json.loads(output)
+
+    result = json.dumps(parsed_data)
```

Comment:
Fixed in d809b460c7, at first I tried to actually wonder why we're returning a raw string instead of an object that Celery should handle fine but changing that is a larger undertaking…